packages/eas-cli/src/utils/queries.ts

```diff
@@ -0,0 +1,107 @@
+import { confirmAsync, selectAsync } from '../prompts';
+import uniqBy from './expodash/uniqBy';
+
+export type PaginatedQueryResponse<QueryReturnType extends Record<string, any>> = {
+  queryResponse: QueryReturnType[];
+  queryResponseRawLength: number;
+};
+
+export type PromptSelectionListItem = {
+  title: string;
+  value: string;
+};
+
+export type PaginatedQueryArguments<QueryReturnType extends Record<string, any>> = {
+  pageSize: number;
+  offset: number;
+  queryToPerform: (
+    pageSize: number,
+    offset: number
+  ) => Promise<PaginatedQueryResponse<QueryReturnType>>;
+  promptOptions: PaginatedQueryPromptOptions<QueryReturnType>;
+};
+
+export interface PaginatedQuerySelectPrompt<QueryReturnType extends Record<string, any>> {
+  readonly title: string;
+  readonly type: 'select';
+  createDisplayTextForSelectionPromptListItem: (queryItem: QueryReturnType) => string;
+  getIdentifierForQueryItem: (queryItem: QueryReturnType) => string;
+}
+
+export interface PaginatedQueryConfirmPrompt<QueryReturnType extends Record<string, any>> {
+  readonly type: 'confirm';
+  readonly title: string;
+  renderListItems: (currentPage: QueryReturnType[]) => void;
+}
+
+type PaginatedQueryPromptOptions<QueryReturnType extends Record<string, any>> =
+  | PaginatedQueryConfirmPrompt<QueryReturnType>
+  | PaginatedQuerySelectPrompt<QueryReturnType>;
+
+/**
+ * Return an empty array when the promptOptions type is PaginatedQueryConfirmPrompt
+ * Returns an array of the selected items when the promptOptions type is PaginatedQuerySelectPrompt
+ */
+export async function performPaginatedQueryAsync<QueryReturnType extends Record<string, any>>(
+  { pageSize, offset, queryToPerform, promptOptions }: PaginatedQueryArguments<QueryReturnType>,
+  accumulator?: QueryReturnType[]
+): Promise<QueryReturnType[]> {
```

Comment:
Spent quite a bit of time trying to get a [conditional return type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html) working here where if promptOptions['type'] was confirm we return void, otherwise return QueryReturnType, but I couldn't quite get it functioning. Would love to see that behavior if anyone wants to take a crack.